File: 29-divide-two-integers/29-divide-two-integers.py
class Solution:
    def divide(self, dividend: int, divisor: int) -> int:
        MAX_INT = 2 ** 31 - 1
        MIN_INT = - 2 ** 31 
        sym = 2
        sign = 1
        if (dividend == MIN_INT and divisor == -1):
            return MAX_INT
        if (dividend > 0):
            sym = sym - 1
            dividend = - dividend
        if (divisor > 0):
            sym = sym - 1
            divisor = - divisor
        
        if (sym == 1):
            sign = -1
        def divide_core(dividend, divisor):
            ans = 0
            while(dividend <= divisor):
                value = divisor
                temp = 1
                while(value >= - 2 ** 30 and value + value >= dividend):
                    value = value << 1
                    temp = temp << 1
                ans += temp
                dividend = dividend - value
            return ans
        res = divide_core(dividend, divisor)
        return sign * res
        
        

File: 29-divide-two-integers/29-divide-two-integers.java
class Solution {
    public int divide(int dividend, int divisor) {
        int sign = 1;
        if(dividend < 0) sign =-sign;
        if(divisor < 0) sign = -sign;
        
        if(dividend > 0)dividend = -dividend;
        if(divisor > 0) divisor = -divisor;
        
        if(sign == 1 && dividend == Integer.MIN_VALUE && divisor == -1)return Integer.MAX_VALUE;
        int ans = this.helper(dividend, divisor);
        return ans * sign;
        
    }
    public int helper(int dividend, int divisor){
        int ans = 0;
        int temp_divisor = 0;
        int quotient = 0;
        while(dividend <= divisor){
            temp_divisor = divisor;
            quotient = 0;
            while(temp_divisor + temp_divisor >= dividend&& temp_divisor >= 0xc0000000){
                temp_divisor += temp_divisor;
                quotient += 1;
            }
            
            dividend -= temp_divisor;
            ans += (1<<quotient);
        }
        
        return ans;
        
        
        
        
    }

    
}